frameworks/keyed/vanillajs-2/src/Main.js

```diff
@@ -0,0 +1,205 @@
+    }
+    removeAllRows() {
+        // ~258 msecs
+        // for(let i=this.rows.length-1;i>=0;i--) {
+        //     tbody.removeChild(this.rows[i]);
+        // }
+        // ~251 msecs
+        // for(let i=0;i<this.rows.length;i++) {
+        //     tbody.removeChild(this.rows[i]);
+        // }
+        // ~216 msecs
+        // var cNode = tbody.cloneNode(false);
+        // tbody.parentNode.replaceChild(cNode ,tbody);
+        // ~212 msecs
+        this.tbody.textContent = "";
+
+        // ~236 msecs
+        // var rangeObj = new Range();
+        // rangeObj.selectNodeContents(tbody);
+        // rangeObj.deleteContents();
+        // ~260 msecs
+        // var last;
+        // while (last = tbody.lastChild) tbody.removeChild(last);
+    }
+    runLots() {
+        this.removeAllRows();
+        this.data = [];
+        this.appendRows(buildData(10000));
+        this.unselect();
+    }
+    clear() {
+        this.data = [];
+        // This is actually a bit faster, but close to cheating
+        // requestAnimationFrame(() => {
+            this.removeAllRows();
+            this.unselect();
+        // });
+    }
+    swapRows() {
+        if (this.data.length > 998) {
+
+            let tmp = this.data[998];
+            this.data[998] = this.data[1];
+            this.data[1] = tmp;
+
+            let a = this.tbody.lastChild.previousSibling,
+            b = a.nextSibling,
+            c = this.tbody.firstChild.nextSibling;
+            this.tbody.insertBefore(this.tbody.replaceChild(a, c), b);
+        }
+
+        // let old_selection = this.store.selected;
+        // this.store.swapRows();
+        // this.updateRows();
+        // this.unselect();
+        // if (old_selection>=0) {
+        //     let idx = this.store.data.findIndex(d => d.id === old_selection);
+        //     if (idx > 0) {
+        //         this.store.select(this.data[idx].id);
+        //         this.selectedRow = this.rows[idx];
+        //         this.selectedRow.className = "danger";
+        //     }
+        // }
+    }
+    appendRows(newData) {
+        // Using a document fragment is slower...
+        // var docfrag = document.createDocumentFragment();
+        // for(let i=this.rows.length;i<this.store.data.length; i++) {
+        //     let tr = this.createRow(this.store.data[i]);
+        //     this.rows[i] = tr;
+        //     this.data[i] = this.store.data[i];
+        //     docfrag.appendChild(tr);
+        // }
+        // this.tbody.appendChild(docfrag);
+
+        // ... than adding directly
+        var tbody = this.tbody;
+        for(let i=0, len=newData.length;i<len; i++) {
+            tbody.appendChild(this.createRow(newData[i]));
+            this.data.push(newData[i]);
+        }
+    }
+    createRow(data) {
+        const tr = rowTemplate.cloneNode(true),
+            td1 = tr.firstChild,
+            a2 = td1.nextSibling.firstChild;
+        tr.data_id = data.id;
+        td1.textContent = data.id;
```

Comment:
It's much faster to use `firstChild.nodeValue` than using `textContent`. If you apply the change to the template then a text node will exist.